### PR TITLE
Use yamllint to check YAML source

### PR DESCRIPTION
**Pull request by adrienverge**
The previous custom script used to check YAML syntax errors only. This
new tool (yamllint) does that too, but also looks for weirdnesses like
key repetition and cosmetic problems such as lines length, trailing
spaces, indentation, etc.

It is configurable via rules and error levels. See
http://yamllint.readthedocs.org/en/latest/ for more details.

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -31,14 +31,13 @@
...
         'scripts/custom_linters/ini_linter.py',
-        'scripts/custom_linters/yaml_linter.py',
         'scripts/custom_linters/jpegtran-linter.sh',
         'scripts/custom_linters/optipng-linter.sh',
         'scripts/custom_linters/pngcrush-linter.sh',
         'scripts/custom_linters/tidy-wrapper.sh',
     ],
     install_requires=[
-        'pyyaml',
+        'yamllint',
```
Comment (sk-) [inline setup.py]: Please move this to the section below (after the comment), where I have all the linters that are pip installable.

File: setup.py [hunk 2]
```diff
@@ -31,14 +31,13 @@
...
         'scripts/pre-commit.hg-lint.sh',
         'scripts/custom_linters/ini_linter.py',
-        'scripts/custom_linters/yaml_linter.py',
         'scripts/custom_linters/jpegtran-linter.sh',
         'scripts/custom_linters/optipng-linter.sh',
         'scripts/custom_linters/pngcrush-linter.sh',
         'scripts/custom_linters/tidy-wrapper.sh',
     ],
     install_requires=[
-        'pyyaml',
```
Comment (sk-) [inline setup.py]: Please keep pyyaml, as it is a direct dependency of git-lint.

File: gitlint/configs/config.yaml [hunk 3]
```diff
@@ -180,9 +180,12 @@ yaml:
   extensions:
   - .yaml
   - .yml
-  command: yaml_linter.py
+  command: yamllint
+  arguments:
+  - "--config-data"
+  - "relaxed"
   filter: "(?P<message>.+)$"
```
Comment (sk-) [inline gitlint/configs/config.yaml]: For stylistic messages you should be matching using the line number as well, so you don;t see the messages of lines you didn't change. For structural errors (invalid file) you don't care about the line number, so you always get the error.

**Comment by sk-** [comment]
Hi Adrien, thanks for the PR. Just some minor issues.

Ans also, please feel free to add yourself to the list of contributors in the README.rst file.

File: gitlint/configs/config.yaml (hunk 3 above)
Comment (adrienverge) [inline gitlint/configs/config.yaml]: I didn't know that, thanks!

File: setup.py (hunk 1 above)
Comment (adrienverge) [inline setup.py]: Sure. Will restore `pyyaml`.

**Comment by adrienverge** [comment]
@sk- Thanks for this fast answer! I updated the pull request.

**Comment by sk-** [comment]
Hi Adrien, thanks for addressing the comments.

But the tests are failing, and I found the reason. The reported lines are off by one.

Take the following yaml file (foo.yaml)
```yaml
a: "b
```

```bash
$ yamllint foo.yml --config-data relaxed
foo.yml
  2:1       error    syntax error: found unexpected end of stream
```

But the offending line is number 1.


Also, I think it would be better if we had a regex like the following:

```
^{filename}:(?P<line>{lines}|\d+(?=\d+: \[error\] syntax error:)):(?P<column>\d+): \[(?P<severity>\S+)\] (?P<message>.+)$
```

In that way, al syntax errors will be shown regarding of whether those lines were modified or not.

**Comment by adrienverge** [comment]
Hi Sebastian,

Thanks, I didn't know tests were failing!

> The reported lines are off by one.

Actually no (other errors are reported on their correct line numbers), but the "end of stream" case is kind of specific. It's because the `pyyaml` library reports the `StreamEndToken` on line 2, which makes sense since your test file does contain `\n` at the end of line 1.

Anyway, using a regex is a great idea. I adapted the one you proposed, to match either:
- syntax errors, on any line
- other errors, on modified lines only

By the way, I wanted to add other tests (for instance checking for key duplicates in YAML), but it seems hard to add a test case in `test/e2etest/data/yaml/` without breaking the syntax error one. Do you have any thoughts?

Thanks again,

Adrien

**Comment by sk-** [comment]
Yes the end to end tests are limited in the sense that it only checks that a new error is triggered or not triggered.

One comment though, I prefer the regex I wrote above, as it captures the line, column and severity even in the case of syntax error.

**Comment by adrienverge** [comment]
> I prefer the regex I wrote above, as it captures the line, column and severity even in the case of syntax error.

Actually your regex did not work, because in the syntax error case it tried to capture the severity while it was already "consumed" in `\[error\] syntax error`.

To make this clearer, here are examples of errors:

    foo.yml:1:4: [warning] trailing spaces  (trailing-spaces)
    foo.yml:5:4: [error] duplication of key "a" in mapping  (key-duplicates)
    foo.yml:6:1: [error] syntax error: found unexpected end of stream

So the only way I see to distinguish syntax errors from other errors, while capturing column and severity is a regex like this (extra spaces are for readability):

    "^{filename}:
     (
         \\d+:(?P<column>\\d+): \\[(?P<severity>\\S+)\\] syntax error:\
       |
         (?P<line>{lines}):(?P<column>\\d+): \\[(?P<severity>\\S+)\\]
     )
     (?P<message>.+)$"

But Python `re` module doesn't accept it because it redefines groups `column` and `severity`.

**Comment by sk-** [comment]
Hi Adrien, the regex works, the problem is that it was missing a colon. The key part of the regex is the lookahead `(?=)`, that checks that the texts follows, without consuming it.

below is an example:
```python
re.match('^.*:(?P<line>1|2|\d+(?=:\d+: \[error\] syntax error:)):(?P<column>\d+): \[(?P<severity>\S+)\] (?P<message>.+)$', 'foo.yml:6:1: [error] syntax error: found unexpected end of stream').groups()
```

**Comment by adrienverge** [comment]
Sebastian,

I'm sorry but I can't get your regex work on both types of errors:

    foo.yml:6:1: [error] syntax error: found unexpected end of stream
    foo.yml:5:4: [error] any other error

Replacing the syntax error by the other one in your python snippet fails. I've tried many combinations but I can't find something better than the regex in my commit.

Could you please show me a working example of what you want by adding a commit on top of mine on a temp branch?

Thanks,

Adrien

**Comment by sk-** [comment]
Hi Adrien,

try first the regex in your command line to see they are working (as shown by the examples), and then you need to escape it when writing it in the YML file.

Below is the same regex applied to the other type of error message:

```python
In [13]: re.match('^.*:(?P<line>1|2|\d+(?=:\d+: \[error\] syntax error:)):(?P<column>\d+): \[(?P<severity>\S+)\] (?P<message>.+)$', 'foo.yml:2:4: [error] any other error').groups()
Out[13]: ('2', '4', 'error', 'any other error')
```
Note that in the regex the fields `lines` and `filename` were already replaced.

The regex, before replacement would be:

```python
r'^{filename}:(?P<line>{lines}|\d+(?=:\d+: \[error\] syntax error:)):(?P<column>\d+): \[(?P<severity>\S+)\] (?P<message>.+)$'
```

Note that this regex would match any syntax errors, regardless of the line number in which they occur, and would also match those errors occurring in the modified lines (`{lines}` pattern).

Let me know if it is clear.

**Comment by sk-** [comment]
Thanks Adrien, I merged the PR. It will take me a couple of days to submit the new version to pypi.
